src/core/server/http/http_server.test.ts

```diff
@@ -1431,3 +1434,71 @@ describe('setup contract', () => {
     });
   });
 });
+
+describe('Graceful shutdown', () => {
+  test('it should wait until the ongoing requests are resolved and reject any new ones', async () => {
+    const gracefulShutdownTimeout = config.gracefulShutdownTimeout.asMilliseconds();
+
+    const { registerRouter, server: innerServer } = await server.setup(config);
+
+    const router = new Router('', logger, enhanceWithContext);
+    router.post(
+      {
+        path: '/',
+        validate: { body: schema.object({ test: schema.number() }) },
+        options: { body: { accepts: 'application/json' } },
+      },
+      async (context, req, res) => {
+        // It takes to resolve the same period of the gracefulShutdownTimeout.
+        // Since we'll trigger the stop a few ms after, it should have time to finish
+        await new Promise((resolve) => setTimeout(resolve, gracefulShutdownTimeout));
+        return res.ok({ body: req.route });
+      }
+    );
+    registerRouter(router);
+
+    await server.start();
+
+    const makeRequest = () => supertest(innerServer.listener).post('/').send({ test: 1 });
+
+    const [firstResponse, , secondResponse] = await Promise.all([
+      // Trigger a request that should succeed
+      makeRequest().expect(200, {
```

Comment:
Is it possible to separate a preparatory phase from the assertion phase to follow [given-when-then](https://martinfowler.com/bliki/GivenWhenThen.html) pattern? you can make these assertions below (as you already do)